Add ERC: Recurring Subscription NFT (#1217)
ethereum/ERCs: ERCS/erc-8027.md

---
eip: 8027
title: Recurring Subscription NFT
description: Manage recurring subscription NFTs via payment compatible with any token approval methods.
author: ant (@0xdevant), cygaar (@cygaar)
discussions-to: https://ethereum-magicians.org/t/erc-8027-recurring-subscription-nft/25482
status: Draft
type: Standards Track
category: ERC
created: 2025-09-16
requires: 165, 721
---

## Abstract

> Recurring Subscription NFT is an extension of [ERC-721](./eip-721.md) token that enables recurring subscription feature by charging the NFT holder [ERC-20](./erc-20.md) tokens via any token approval methods.

Reviewer: Link target should be `./eip-N.md` in this repo:
```suggestion
Recurring Subscription NFT is an extension of [ERC-721](./eip-721.md) token that enables recurring subscription feature by charging the NFT holder [ERC-20](./eip-20.md) tokens via any token approval methods.
```

The interface suggests specifications that allow users to renew/extend subscriptions, service provider to charge users automatically for recurring subscription by their subscription data and [EIP-712](./eip-712.md) signature. Also an optional function for users to cancel their subscription by revoking the respective token approval.

## Motivation

NFTs are commonly used as ownership verification on decentralized apps or membership passes to communities, events, and more. However, these use cases often fall into either being a token of ownership that have no expiration dates, or a one-off ticket for verifying identity/reservation thus no recurring payments involved for both cases. However for many real-world applications with paid subscriptions, they would prefer a middle ground - to keep an account or membership valid until the user stops paying for the subscription.

Currently there exists no standard implementation for subscription service that allows:

1. service provider to easily configure a subscription

> 2. users to setup a seamless payment flow via gasless pre-approval for any [ERC-20](./erc-20.md) with an expiry date

Reviewer: Link form:
```suggestion
2. users to setup a seamless payment flow via gasless pre-approval for any [ERC-20](./eip-20.md) with an expiry date
```

3. the above while restricting the service provider from only automatically charging by each cycle instead of a full amount to keep users' subscriptions active

On the other hand a common interface will make it easier for future projects to develop subscription-based NFTs. In the current Web2 world, it's hard for a user to see or manage all of their subscriptions in one place. With a common standard for subscriptions, it will be easy for a single application to determine the number of subscriptions a user has, see when they expire, and renew/cancel them as requested.

> Overall this standard offers a more efficient approach for both users and the platform. Rather than forcing users to pay a huge lump sum to tie them into a long subscription plan, or having users locked a fixed amount of funds in advance waiting to be charged each month, users can just spend their funds as they please, and keep enough funds to be deducted by the service provider for the next cycle of subscription without going through a manual subscription process - basically like a debit card for recurring subscription, and more similar to how an auto subscription would work in web2.

Reviewer: `locked` -> `lock`, and `web2` -> `Web2` for consistency with line 30:
```suggestion
Overall this standard offers a more efficient approach for both users and the platform. Rather than forcing users to pay a huge lump sum to tie them into a long subscription plan, or having users lock a fixed amount of funds in advance waiting to be charged each month, users can just spend their funds as they please, and keep enough funds to be deducted by the service provider for the next cycle of subscription without going through a manual subscription process - basically like a debit card for recurring subscription, and more similar to how an auto subscription would work in Web2.
```

## Specification

> The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Reviewer: Link the RFCs (rfc-editor.org), add RFC 8174. RFC 8174 clarifies RFC 2119, as RFC 2119 alone can cause confusion over the lowercase variant of the keywords.
Also changes `key words` -> `keywords`.
```suggestion
The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).
```

     */
    function getSubscriptionConfig() external view returns (SubscriptionConfig memory);
}
```

The `isRenewable(uint256 tokenId)` function MAY be implemented as `pure` or `view`.

The `expiresAt(uint256 tokenId)` function MAY be implemented as `pure` or `view`.

The `getRenewalPrice(uint256 tokenId)` function MAY be implemented as `pure` or `view`.

The `renewSubscription(uint256 tokenId, uint128 planIdx, uint64 numOfIntervals)` function MAY be implemented as `external` or `public`.

The `chargeRecurringSubscription(RecurringSubscriptionData recurringSubscriptionData)` function MAY be implemented as `external` or `public`.

The `SubscriptionExtended` event MUST be emitted whenever a subscription is extended.

The `RecurringSubscriptionCharged` event MUST be emitted whenever the service provider charges a user for a recurring subscription.

The `supportsInterface` method MUST return `true` when called with `0xe6997336`.

## Rationale

This standard aims to make on-chain recurring subscriptions as generic and as easy to integrate with as possible by having minimal configurations. Here are few design choices being made to fulfill these purposes.

> 1. The `SubscriptionConfig` allows multiple plans to be configured - the standard supports multiple plans to enable tierd subscription plans and only using [ERC-20](./erc-20.md) as payment token will enable recurring subscription due to its approval feature. The NFT itself represents ownership of a subscription and there is no facilitation of how the NFT should be minted or transferred.

Reviewer: `tierd` → `tiered`, and link form:
```suggestion
1. The `SubscriptionConfig` allows multiple plans to be configured - the standard supports multiple plans to enable tiered subscription plans and only using [ERC-20](./eip-20.md) as payment token will enable recurring subscription due to its approval feature. The NFT itself represents ownership of a subscription and there is no facilitation of how the NFT should be minted or transferred.
```

> 2. `RecurringSubscriptionData` passed in `chargeRecurringSubscription` is designed to be generic in order to integrate with any [ERC-20](./erc-20.md) approval methods:

Reviewer: Link form:
```suggestion
2. `RecurringSubscriptionData` passed in `chargeRecurringSubscription` is designed to be generic in order to integrate with any [ERC-20](./eip-20.md) approval methods:
```

   ```solidity
   struct RecurringSubscriptionData {
       uint256 tokenId;
       uint128 planIdx;
       uint64 numOfIntervals;
       bytes tokenApprovalData;
       bytes extraVerificationData;
   }
   ```

>    `tokenid`: Token id of the Subscription NFT

Reviewer: Casing: the field is `tokenId`:
```suggestion
   `tokenId`: Token id of the Subscription NFT
```

   `planIdx`: Index of the plan in `planPrices` array

   `numOfIntervals`: The number of `billingInterval` for verifying total pre-approval amount of the recurring subscription

>    `tokenApprovalData` Encoded data to handle different token approval logic such as Permit/Permit2

Reviewer: It is not clear what Permit/Permit2 is, please define it such that the reader knows what is meant/intended by this.

   `extraVerificationData` Encoded data to handle extra verification logic for example a signed [EIP-712](./eip-712.md) signature if more verifications are needed.

3. For actual implementation in `ERC8027.sol`, customizations on:

>    - verifying token approval and charging automatically should be done by overriding `_verifyApprovalAndCharge` to integrate with any common [ERC-20](./erc-20.md) approval methods e.g. [ERC-2612](./erc-2612.md), [ERC-3009](./erc-3009.md) and Permit2

Reviewer: Link form for the referenced ERCs:
```suggestion
   - verifying token approval and charging automatically should be done by overriding `_verifyApprovalAndCharge` to integrate with any common [ERC-20](./eip-20.md) approval methods e.g. [ERC-2612](./eip-2612.md), [ERC-3009](./eip-3009.md) and Permit2
```

>    - verifing inputs passed from the function caller should be done by overriding `_verifyInputs`

Reviewer: `verifing` → `verifying`:
```suggestion
   - verifying inputs passed from the function caller should be done by overriding `_verifyInputs`
```

   - paying for manual subscription should be done by overriding `_pay`
   - extending subscription should be done by overriding `_extendSubscription`

### Considerations on token pre-approval via Account Abstraction (AA)

> Even though we could somehow achieve something similar to gasless pre-approval with the help of [EIP-7702](./eip-7702.md), user’s EOA will need to delegate to an external contract to act as an smart account which would impose extra security vectors that integrators need to consider, and the subscription feature is still lacking nonetheless.

Reviewer: `an smart` → `a smart`, expand EOA on first use, and use a straight apostrophe:
```suggestion
Even though we could somehow achieve something similar to gasless pre-approval with the help of [EIP-7702](./eip-7702.md), the user's Externally Owned Account (EOA) will need to delegate to an external contract to act as a smart account which would impose extra security vectors that integrators need to consider, and the subscription feature is still lacking nonetheless.
```

> More importantly this interface is in fact compatitable with such smart accounts as long as they are able to receive NFT i.e. with `onERC721Received` implemented thus this interface provides a standardized framework while being AA-compatible for apps to offer subscription features in a more composable way.

Reviewer: `compatitable` → `compatible`:
```suggestion
More importantly this interface is in fact compatible with such smart accounts as long as they are able to receive NFT i.e. with `onERC721Received` implemented thus this interface provides a standardized framework while being AA-compatible for apps to offer subscription features in a more composable way.
```

### Subscription Management

> - **Manual subscription:** Users should be able to renew their subscriptions by directly transferring either native or [ERC-20](./erc-20.md) tokens to the service provider thus the `renewSubscription` function. Users will specify the index of subscription plan i.e. `planIdx` and the number of interval to subscribe for i.e. `numOfIntervals`.

Reviewer: Link form:
```suggestion
- **Manual subscription:** Users should be able to renew their subscriptions by directly transferring either native or [ERC-20](./eip-20.md) tokens to the service provider thus the `renewSubscription` function. Users will specify the index of subscription plan i.e. `planIdx` and the number of interval to subscribe for i.e. `numOfIntervals`.
```

- `expiresAt` function allows users and applications to directly confirm the validity of a Subscription NFT by checking its expiration date, and `getSubscriptionDetails` function will get both the expiration date and the subscription plan the Subscription NFT belongs to.

- `getRenewalPrice` helps users and applications to calculate the price of the subscription given the plan it belongs and the number of intervals the subscription will continue for.

- `isRenewable` function gives users and applications the information whether a subscription for a certain NFT or all NFTs could be renewed once expired.

> - `getSubscriptionConfig` will give users and applications the information about the configuration of the subscription such as which payment token, which service provider will receive the subscription fee, how long is each interval is, and the price for each plan.

Reviewer: Doubled `is`:
```suggestion
- `getSubscriptionConfig` will give users and applications the information about the configuration of the subscription such as which payment token, which service provider will receive the subscription fee, how long each interval is, and the price for each plan.
```

> - Finally it's important to know that only using [ERC-20](./erc-20.md) as payment token can enable recurring subscription as we cannot directly transfer native token on behalf of users without introducing external dependencies.

Reviewer: Link form:
```suggestion
- Finally it's important to know that only using [ERC-20](./eip-20.md) as payment token can enable recurring subscription as we cannot directly transfer native token on behalf of users without introducing external dependencies.
```

## Backwards Compatibility

> This standard can be fully [ERC-721](./erc-721.md) compatible by adding an extension function set, and payment can be integrated with any [ERC-20](./erc-20.md) out of the box.

Reviewer: Link form:
```suggestion
This standard can be fully [ERC-721](./eip-721.md) compatible by adding an extension function set, and payment can be integrated with any [ERC-20](./eip-20.md) out of the box.
```

> This standard is also fully compatible with smart accounts that utilize Account Abstraction standards such as [EIP-4337](./erc-4337.md) and [EIP-7702](./erc-7702.md) as long as they are able to receive NFT i.e. with `onERC721Received` implemented or with `ERC721Holder.sol` from OpenZeppelin inherited.

Reviewer: `EIP-4337` → `ERC-4337` (category ERC), and link forms:
```suggestion
This standard is also fully compatible with smart accounts that utilize Account Abstraction standards such as [ERC-4337](./eip-4337.md) and [EIP-7702](./eip-7702.md) as long as they are able to receive NFT i.e. with `onERC721Received` implemented or with `ERC721Holder.sol` from OpenZeppelin inherited.
```
